src/components/annual-report/report-data/section4/useFetchSectionFourData.js

```diff
@@ -0,0 +1,111 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useState } from 'react'
+
+const section4ByLevel = {
+    data_detail_by_level: {
+        resource: 'analytics.json',
+        params: ({ dataElements, orgUnits, orgUnitLevel, periods }) => ({
+            dimension: `dx:${dataElements.join(';')},ou:${
+                orgUnits.join(';') + ';' + orgUnitLevel
+            },pe:${periods.join(';')}`,
+        }),
+    },
+}
+
+const section4Overall = {
+    data_over_all_org_units: {
+        resource: 'analytics.json',
+        params: ({ dataElements, orgUnits, periods }) => ({
+            dimension: `dx:${dataElements.join(';')},ou:${orgUnits.join(
+                ';'
+            )},pe:${periods.join(';')}`,
+        }),
+    },
+}
+
+export const fetchDataByLevel = async ({ engine, variables }) => {
+    const dataBySubPeriod = await engine.query(section4ByLevel, {
+        variables,
+    })
+    return dataBySubPeriod
+}
+
+export const useFetchSectionFourData = () => {
+    const [loading, setLoading] = useState(false)
+    const [data, setData] = useState(null)
+    const [error, setError] = useState(null)
+
+    const engine = useDataEngine()
+
+    const refetch = async ({ variables = {} }) => {
```

Comment:
I saw a comment in SectionFour.js that you want this to be stable -- you could wrap this in `useCallback` with `engine` as a dependency:
```js
const refetch = useCallback(async ({ variables = {} }) => {
    ...
}, [engine])
```